src/main/java/org/opentripplanner/street/model/vertex/BarrierVertex.java

```diff
@@ -33,4 +32,32 @@ public StreetTraversalPermission getBarrierPermissions() {
   public void setBarrierPermissions(StreetTraversalPermission barrierPermissions) {
     this.barrierPermissions = barrierPermissions;
   }
+
+  /*
+   * Barrier vertex at the end of a way does not make sense, because
+   * it creates discontinuity of routing in a single point.
+   * This method examines if traversal limitations can be removed.
+   * The logic examines edges referring to the vertex, so it should be
+   * applied only after the vertex has been linked to the graph.
+   */
+  public void makeBarrierAtEndReachable() {
+    var edgeCount = this.getDegreeOut() + this.getDegreeIn();
+    var needsFix = false;
+    if (edgeCount == 1) {
+      needsFix = true;
+    } else if (edgeCount == 2) {
+      var out = this.getOutgoing();
+      var in = this.getIncoming();
+      if (
+        out.isEmpty() ||
+        in.isEmpty() ||
+        out.iterator().next().getToVertex() == in.iterator().next().getFromVertex()
+      ) {
+        needsFix = true;
+      }
```

Comment:
I think we could add a comment here what this case handles as this is not as obvious.